// ironworks/src/workers/HttpWorker.js
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var path = require('path');
var _ = require('lodash');
var hapi = require('hapi');
var idHelper = require('..//helpers/idHelper');
var CommEmit = require('..//eventing/CommEmit');
var Worker = require('../../src/workers/Worker');
var HttpWorker = (function (_super) {
    __extends(HttpWorker, _super);
    function HttpWorker(comm, service, opts) {
        _super.call(this, [], comm, service, {
            id: idHelper.newId(),
            name: 'iw-http'
        });
        var defOpts = {
            port: process.env.VCAP_APP_PORT,
            rootSitePagePath: 'index.html'
        };
        this.opts = this.opts.beAdoptedBy(defOpts, 'worker');
        this.opts.merge(opts);
        console.log('Finished contructing HttpWorker');
    }
    HttpWorker.prototype.init = function (callback) {
        var instance = this;
        this.httpServer = new hapi.Server(this.opts.get('hapi'));
        this.httpServer.connection({
            port: this.opts.get('port')
        });
        var serveApi = this.opts.has('apiUri');
        if (serveApi) {
            var api = this.opts.get('apiUri');
            if (api[0] !== '/') {
                api = '/' + api;
            }
            if (api[api.length - 1] === '/') {
                api = api.substring(0, api.length - 1);
            }
            this.httpServer.route({
                method: 'POST',
                path: api + '/{prefix}/{service}/{method}/{worker}/{event}',
                config: {
                    payload: {
                        allow: 'application/json',
                        parse: true
                    }
                },
                handler: function (req, reply) {
                    console.log('POST CALLED');
                    console.log(JSON.stringify(req.paramsArray));
                    var event = new CommEmit({
                        id: idHelper.newId(),
                        emitter: {
                            id: _.isUndefined(req.query.myId) ? instance.me.id : req.query.myId,
                            name: _.isUndefined(req.query.myName) ? instance.me.name : req.query.myName
                        },
                        prefix: req.paramsArray[0],
                        service: req.paramsArray[1],
                        method: req.paramsArray[2],
                        worker: req.paramsArray[3],
                        name: req.paramsArray[4]
                    });
                    switch (event.method) {
                        case 'tell':
                            if (req.method !== 'post') {
                                return;
                            }
                            instance.tell(event);
                            reply(200);
                            break;
                        case 'inform':
                            if (req.method !== 'post') {
                                return;
                            }
                            instance.inform(event, req.payload);
                            reply(200);
                            break;
                        case 'confirm':
                            instance.confirm(event, function (e) {
                                reply(e === null ? 200 : e);
                            });
                            break;
                        case 'ask':
                            instance.ask(event, function (e, answer) {
                                reply(e, answer);
                            });
                            break;
                        case 'request':
                            instance.request(event, req.payload, function (e, response) {
                                reply(e, response);
                            });
                            break;
                    }
                }
            });
            this.httpServer.route({
                method: 'GET',
                path: api + '/{prefix}/{service}/{method}/{worker}/{event}',
                handler: function (req, reply) {
                    console.log('GET CALLED');
                    console.log(JSON.stringify(req.paramsArray));
                    var event = new CommEmit({
                        id: idHelper.newId(),
                        emitter: {
                            id: _.isUndefined(req.query.myId) ? instance.me.id : req.query.myId,
                            name: _.isUndefined(req.query.myName) ? instance.me.name : req.query.myName
                        },
                        prefix: req.paramsArray[0],
                        service: req.paramsArray[1],
                        method: req.paramsArray[2],
                        worker: req.paramsArray[3],
                        name: req.paramsArray[4]
                    });
                    switch (event.method) {
                        case 'tell':
                            if (req.method !== 'post') {
                                return;
                            }
                            instance.tell(event);
                            reply(200);
                            break;
                        case 'inform':
                            if (req.method !== 'post') {
                                return;
                            }
                            instance.inform(event, req.payload);
                            reply(200);
                            break;
                        case 'confirm':
                            instance.confirm(event, function (e) {
                                reply(e === null ? 200 : e);
                            });
                            break;
                        case 'ask':
                            instance.ask(event, function (e, answer) {
                                reply(e, answer);
                            });
                            break;
                        case 'request':
                            instance.request(event, req.payload, function (e, response) {
                                reply(e, response);
                            });
                            break;
                    }
                }
            });
        }
        var servePublicRoot = this.opts.has('hapi.connections.routes.files.relativeTo');
        if (servePublicRoot) {
            this.httpServer.route({
                method: 'GET',
                path: '/{p*}',
                handler: function (req, reply) {
                    console.log('FILE ROUTER CALLED');
                    console.log(JSON.stringify(req.paramsArray));
                    if (req.path === '/') {
                        req.path = instance.opts.get('rootSitePagePath');
                    }
                    var root = instance.opts.get('hapi.connections.routes.files.relativeTo');
                    var filePath = path.join(root, req.path);
                    reply.file(filePath);
                }
            });
        }
        if (!_.isUndefined(callback)) {
            callback(null);
        }
    };
    HttpWorker.prototype.start = function (dependencies, callback) {
        var _this = this;
        this.httpServer.start(function (e) {
            if (_.isUndefined(e)) {
                e = null;
            }
            if (e === null) {
                _this.tell('ready');
            }
            else {
                _this.inform('error', e);
            }
            if (!_.isUndefined(callback)) {
                process.nextTick(function () {
                    callback(e);
                });
            }
        });
    };
    HttpWorker.prototype.dispose = function (callback) {
        if (!_.isUndefined(this.httpServer)) {
            this.httpServer.stop(void 0, function () {
                if (!_.isUndefined(callback)) {
                    callback();
                }
            });
        }
        else {
            if (!_.isUndefined(callback)) {
                process.nextTick(function () {
                    callback();
                });
            }
        }
    };
    return HttpWorker;
})(Worker);
module.exports = HttpWorker;
